### PR TITLE
Ingest assay layers into separate arrays

### PR DESCRIPTION
**Pull request by aaronwolen**
This updates our storage strategy to ingest each data-containing slot of a Seurat assay (i.e., a *layer*) into a separate array within X. The only user-facing change is that `X` is now a group (`AssayMatrixGroup`) of arrays. 

Other changes:
- `SCGroup$to_seurat_assay()` gains a `layers` argument to specify which layers (arrays within `X`) should be retrieved.

### CODE REVIEW

**Comment by aaronwolen** [comment]
CC @avrobullet

**Review by ckrilow — APPROVED** [review]
These changes look good to me.

File: R/AssayMatrix.R [hunk 1]
```diff
@@ -23,17 +23,23 @@ AssayMatrix <- R6::R6Class(
...
       if (inherits(x, "dgCMatrix")) {
           message("Converting to dgTMatrix")
           x <- as(x, "dgTMatrix")
         }
       stopifnot("'x' must be a dgTMatrix" = inherits(x, "dgTMatrix"))
-      stopifnot(is_scalar_character(attr))
+      if (missing(index_cols)) {
+        stop("Must provide 'index_cols' to name the index columns")
+      }
+      stopifnot(is_scalar_character(value_col))
```
Comment (eddelbuettel) [inline R/AssayMatrix.R:39]: If you want to fancy, or more compact, you can write the three lines as

```r
stopifnot("'x' must be a dgTMatrix" = inherits(x, "dgTMatrix"),
               "'index_cols' must be present" = !missing(index_cols),
               "'value_col' must be scalar" = is_scalar_value(value_col))
```

File: R/AssayMatrix.R [hunk 2]
```diff
@@ -42,7 +48,10 @@ AssayMatrix <- R6::R6Class(
     #' @param index_cols A column index, either numeric with a column index, or
     #' character with a column name, designating one or more index columns. All
     #' other columns are ingested as attributes.
-    from_dataframe = function(x, index_cols = c("var_id", "obs_id")) {
+    from_dataframe = function(x, index_cols) {
+      if (missing(index_cols)) {
+        stop("Must provide 'index_cols' to identify the index columns")
+      }
```
Comment (eddelbuettel) [inline R/AssayMatrix.R:54]: As above

**Review by eddelbuettel — APPROVED** [review]
LGTM modulo a minor stylistic suggestion

File: R/AssayMatrix.R (hunk 1 above)
Comment (aaronwolen) [reply R/AssayMatrix.R:39]: I _always_ want to be fancy :) 

You're right, we've discussed previously and I do prefer that convention.

**Comment by avrobullet** [comment]
Not sure if this is necessary, but I noticed that the list `X` description in `SCGroup.R` has not been updated to say `AssayMatrixGroup`:

```
...
25 - #' @field X named list of [`AssayMatrix`] object containing matrix-like
26 - #' assay data with string dimensions `obs_id` and `var_id` that align to the
27 - #' dimensions of the `obs` and `var` arrays, respectively.
28 - X = list(),
...
```

**Pull request merged**